File: python/avin/domain/bar.py
# ────────────────────────────────────────────────────────────────────────────
#  AVIN
#  Understand the market before trading it.
#
#  https://avin.info
# ────────────────────────────────────────────────────────────────────────────

from __future__ import annotations

from datetime import datetime as DateTime

from avin._native import PyBar
from avin.domain.bar_direction import BarDirection
from avin.domain.price_range import PriceRange


class Bar:
    """
    OHLCV market bar.

    The timestamp identifies the start of the bar.

    Bar values are assumed to contain valid market data and are not
    validated during construction.

    Parameters
    ----------
    ts : int
        Bar start timestamp in Unix nanoseconds.
    o : float
        Open price.
    h : float
        High price.
    l : float
        Low price.
    c : float
        Close price.
    v : int
        Volume.

    Attributes
    ----------
    ts : int
        Bar start timestamp in Unix nanoseconds.
    o : float
        Open price.
    h : float
        High price.
    l : float
        Low price.
    c : float
        Close price.
    v : int
        Volume.

    Examples
    --------
    >>> bar = Bar(
    ...     1787235600000000000,
    ...     100.1,
    ...     105.5,
    ...     98.8,
    ...     103.3,
    ...     5000,
    ... )
    >>> bar.o
    100.1
    >>> bar.c
    103.3
    >>> bar.is_bull()
    True
    >>> print(bar.dt())
    2026-08-20 14:20:00+00:00
    >>> 101.0 in bar
    True
    >>> r = bar.range()
    >>> print(r)
    [98.8, 105.5]
    """

    __slots__ = ("_inner",)

    def __init__(
        self, ts: int, o: float, h: float, l: float, c: float, v: int
    ) -> None:
        self._inner = PyBar(ts, o, h, l, c, v)

    def __str__(self) -> str:
        return self._inner.display()

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Bar):
            return NotImplemented

        return self._inner.eq(other._inner)

    def __contains__(self, price: float) -> bool:
        return self._inner.contains(price)

    @property
    def ts(self) -> int:
        """
        Bar start timestamp in Unix nanoseconds.
        """
        return self._inner.ts()

    @property
    def o(self) -> float:
        """
        Open price.
        """
        return self._inner.o()

    @property
    def h(self) -> float:
        """
        High price.
        """
        return self._inner.h()

    @property
    def l(self) -> float:
        """
        Low price.
        """
        return self._inner.l()

    @property
    def c(self) -> float:
        """
        Close price.
        """
        return self._inner.c()

    @property
    def v(self) -> int:
        """
        Volume.
        """
        return self._inner.v()

    def dt(self) -> DateTime:
        """
        Return the bar start timestamp as a UTC datetime.
        """
        return self._inner.dt()

    def direction(self) -> BarDirection:
        """
        Return the bar direction.
        """
        return BarDirection._from_native(self._inner.direction())

    def is_bear(self) -> bool:
        """
        Return whether the bar is bearish.
        """
        return self._inner.is_bear()

    def is_bull(self) -> bool:
        """
        Return whether the bar is bullish.
        """
        return self._inner.is_bull()

    def is_neutral(self) -> bool:
        """
        Return whether the bar is neutral (open == close).
        """
        return self._inner.is_neutral()

    def range(self) -> PriceRange:
        """
        Return the full price range of the bar, `[L, H]`.
        """
        return PriceRange._from_native(self._inner.range())

    def body(self) -> PriceRange:
        """
        Return the body price range, `[min(O, C), max(O, C)]`.
        """
        return PriceRange._from_native(self._inner.body())

    def lower(self) -> PriceRange:
        """
        Return the lower wick price range, `[L, min(O, C)]`.
        """
        return PriceRange._from_native(self._inner.lower())

    def upper(self) -> PriceRange:
        """
        Return the upper wick price range, `[max(O, C), H]`.
        """
        return PriceRange._from_native(self._inner.upper())

    def contains(self, price: float) -> bool:
        """
        Check whether a price is within the closed bar range `[L, H]`.
        """
        return self._inner.contains(price)

    @classmethod
    def _from_native(cls, native: PyBar) -> Bar:
        obj = cls.__new__(cls)
        obj._inner = native

        return obj

File: crates/avin_domain/src/chart/bar.rs
// ───────────────────────────────────────────────────────────────────────────
// AVIN
// Understand the market before trading it.
//
// https://avin.info
// ───────────────────────────────────────────────────────────────────────────

use std::fmt::Display;

use avin_core::{Price, PriceRange, Time};
use chrono::{DateTime, Utc};

use crate::{BarDirection, DomainError};

/// An OHLCV market bar.
///
/// Bar values are assumed to contain valid market data; bar-level invariants
/// must be validated before constructing bars.
#[derive(Debug, Clone, Copy, PartialEq)]
pub struct Bar {
    /// Bar start time.
    pub time: Time,
    /// Open price.
    pub o: Price,
    /// High price.
    pub h: Price,
    /// Low price.
    pub l: Price,
    /// Close price.
    pub c: Price,
    /// Volume.
    pub v: u64,
}

impl Bar {
    /// Creates a bar from trusted OHLCV values without validation.
    pub fn new(
        time: Time,
        o: Price,
        h: Price,
        l: Price,
        c: Price,
        v: u64,
    ) -> Result<Self, DomainError> {
        if o < l || o > h {
            return Err(DomainError::Bar(format!(
                "open price {o} is outside bar range [{l}, {h}]"
            )));
        }

        if c < l || c > h {
            return Err(DomainError::Bar(format!(
                "close price {c} is outside bar range [{l}, {h}]"
            )));
        }

        Ok(Self::new_unchecked(time, o, h, l, c, v))
    }

    // TODO: вынести в "приватный трейт"
    pub fn new_unchecked(
        time: Time,
        o: Price,
        h: Price,
        l: Price,
        c: Price,
        v: u64,
    ) -> Bar {
        Bar {
            time,
            o,
            h,
            l,
            c,
            v,
        }
    }

    /// Returns the bar start time as a UTC datetime.
    pub fn dt(&self) -> DateTime<Utc> {
        self.time.dt()
    }

    /// Returns the bar direction.
    pub fn direction(&self) -> BarDirection {
        if self.is_bull() {
            BarDirection::Bull
        } else if self.is_bear() {
            BarDirection::Bear
        } else {
            BarDirection::Neutral
        }
    }

    /// Returns whether the bar is bearish.
    pub fn is_bear(&self) -> bool {
        self.o > self.c
    }

    /// Returns whether the bar is bullish.
    pub fn is_bull(&self) -> bool {
        self.o < self.c
    }

    /// Returns whether the bar is neutral.
    pub fn is_neutral(&self) -> bool {
        self.o == self.c
    }

    /// Returns the full price range of the bar: `[L, H]`.
    ///
    /// # Panics
    ///
    /// Panics if the price range cannot be constructed from the bar values.
    pub fn range(&self) -> PriceRange {
        PriceRange::new(self.l, self.h).unwrap()
    }

    /// Returns the body price range: `[min(O, C), max(O, C)]`.
    ///
    /// # Panics
    ///
    /// Panics if the price range cannot be constructed from the bar values.
    pub fn body(&self) -> PriceRange {
        PriceRange::new(self.o.min(self.c), self.o.max(self.c)).unwrap()
    }

    /// Returns the lower wick price range: `[L, min(O, C)]`.
    ///
    /// # Panics
    ///
    /// Panics if the price range cannot be constructed from the bar values.
    pub fn lower(&self) -> PriceRange {
        PriceRange::new(self.l, self.o.min(self.c)).unwrap()
    }

    /// Returns the upper wick price range: `[max(O, C), H]`.
    ///
    /// # Panics
    ///
    /// Panics if the price range cannot be constructed from the bar values.
    pub fn upper(&self) -> PriceRange {
        PriceRange::new(self.o.max(self.c), self.h).unwrap()
    }

    /// Returns whether the closed price range `[L, H]` contains the given
    /// price.
    pub fn contains(&self, price: Price) -> bool {
        self.range().contains(price)
    }
}

impl Display for Bar {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{} O={} H={} L={} C={} V={}",
            self.time, self.o, self.h, self.l, self.c, self.v
        )
    }
}

#[cfg(test)]
mod tests {
    use std::str::FromStr;

    use super::*;

    #[test]
    fn new_and_dt() {
        let time = Time::from_str("2026-08-20 14:20:05").unwrap();
        let open = Price::new(10.0).unwrap();
        let high = Price::new(11.0).unwrap();
        let low = Price::new(9.0).unwrap();
        let close = Price::new(10.5).unwrap();
        let vol = 5000;
        let bar = Bar::new(time, open, high, low, close, vol).unwrap();

        assert_eq!(bar.time, time);
        assert_eq!(bar.o.value(), 10.0);
        assert_eq!(bar.h.value(), 11.0);
        assert_eq!(bar.l.value(), 9.0);
        assert_eq!(bar.c.value(), 10.5);
        assert_eq!(bar.v, 5000);

        let dt = time.dt();
        assert_eq!(bar.dt(), dt);
    }

    #[test]
    fn direction() {
        let time = Time::from_str("2026-08-20 14:20:05").unwrap();
        let open = Price::new(10.0).unwrap();
        let high = Price::new(11.0).unwrap();
        let low = Price::new(9.0).unwrap();
        let close = Price::new(10.5).unwrap();
        let vol = 5000;

        let bull = Bar::new(time, open, high, low, close, vol).unwrap();
        assert!(bull.is_bull());
        assert!(!bull.is_bear());
        assert!(!bull.is_neutral());
        assert_eq!(bull.direction(), BarDirection::Bull);

        let close = Price::new(9.5).unwrap();
        let bear = Bar::new(time, open, high, low, close, vol).unwrap();
        assert!(!bear.is_bull());
        assert!(bear.is_bear());
        assert!(!bear.is_neutral());
        assert_eq!(bear.direction(), BarDirection::Bear);

        let close = Price::new(10.0).unwrap();
        let neutral = Bar::new(time, open, high, low, close, vol).unwrap();
        assert!(!neutral.is_bull());
        assert!(!neutral.is_bear());
        assert!(neutral.is_neutral());
        assert_eq!(neutral.direction(), BarDirection::Neutral);
    }

    #[test]
    fn ranges() {
        let time = Time::from_str("2026-08-20 14:20:05").unwrap();
        let open = Price::new(10.0).unwrap();
        let high = Price::new(11.0).unwrap();
        let low = Price::new(9.0).unwrap();
        let close = Price::new(10.5).unwrap();
        let vol = 5000;

        let bull = Bar::new(time, open, high, low, close, vol).unwrap();
        assert_eq!(bull.range(), PriceRange::new(low, high).unwrap());
        assert_eq!(bull.body(), PriceRange::new(open, close).unwrap());
        assert_eq!(bull.lower(), PriceRange::new(low, open).unwrap());
        assert_eq!(bull.upper(), PriceRange::new(close, high).unwrap());

        let close = Price::new(9.5).unwrap();
        let bear = Bar::new(time, open, high, low, close, vol).unwrap();
        assert_eq!(bear.range(), PriceRange::new(low, high).unwrap());
        assert_eq!(bear.body(), PriceRange::new(close, open).unwrap());
        assert_eq!(bear.lower(), PriceRange::new(low, close).unwrap());
        assert_eq!(bear.upper(), PriceRange::new(open, high).unwrap());

        let close = Price::new(10.0).unwrap();
        let neutral = Bar::new(time, open, high, low, close, vol).unwrap();
        assert_eq!(neutral.range(), PriceRange::new(low, high).unwrap());
        assert_eq!(neutral.body(), PriceRange::new(open, close).unwrap());
        assert_eq!(neutral.body(), PriceRange::new(close, open).unwrap());
        assert_eq!(neutral.lower(), PriceRange::new(low, open).unwrap());
        assert_eq!(neutral.lower(), PriceRange::new(low, close).unwrap());
        assert_eq!(neutral.upper(), PriceRange::new(open, high).unwrap());
        assert_eq!(neutral.upper(), PriceRange::new(close, high).unwrap());
    }

    #[test]
    fn contains() {
        let time = Time::from_str("2026-08-20 14:20:05").unwrap();
        let open = Price::new(10.0).unwrap();
        let high = Price::new(11.0).unwrap();
        let low = Price::new(9.0).unwrap();
        let close = Price::new(10.5).unwrap();
        let vol = 5000;

        let bar = Bar::new(time, open, high, low, close, vol).unwrap();

        assert!(bar.contains(Price::new(10.3).unwrap()));
        assert!(bar.contains(Price::new(9.0).unwrap()));
        assert!(bar.contains(Price::new(11.0).unwrap()));

        assert!(!bar.contains(Price::new(11.01).unwrap()));
        assert!(!bar.contains(Price::new(8.99).unwrap()));
    }

    #[test]
    fn display() {
        let time = Time::from_str("2026-08-20 14:20:05").unwrap();
        let open = Price::new(10.0).unwrap();
        let high = Price::new(11.0).unwrap();
        let low = Price::new(9.0).unwrap();
        let close = Price::new(10.5).unwrap();
        let vol = 5000;

        let bar = Bar::new(time, open, high, low, close, vol).unwrap();

        assert_eq!(
            bar.to_string(),
            "2026-08-20 14:20:05 O=10 H=11 L=9 C=10.5 V=5000"
        );
    }
}
